### PR TITLE
feat: Adding Spanner database provider for retrieval service.

### PR DESCRIPTION
**Pull request by gauravpurohit06**
Adding Spanner database provider for retrieval service

### CODE REVIEW

File: retrieval_service/datastore/providers/spanner.py [hunk 1]
```diff
@@ -0,0 +1,692 @@
...
+            flight_number (str): The flight number.
+            departure_airport (str): The departure airport code.
+            arrival_airport (str): The arrival airport code.
+            departure_time (str): The departure time of the flight.
+            arrival_time (str): The arrival time of the flight.
+
+        Raises:
+            NotImplementedError: This method is not implemented.
+        """
+        raise NotImplementedError("Not Implemented")
```
Comment (Yuan325) [inline retrieval_service/datastore/providers/spanner.py:623]: Implement this as well?

File: retrieval_service/datastore/providers/spanner.py [hunk 2]
```diff
@@ -0,0 +1,692 @@
...
+
+        This method is not implemented.
+
+        Args:
+            user_id (str): The ID of the user.
+
+        Raises:
+            NotImplementedError: This method is not implemented.
+        """
+        raise NotImplementedError("Not Implemented")
```
Comment (Yuan325) [inline retrieval_service/datastore/providers/spanner.py:640]: this too

File: retrieval_service/datastore/providers/spanner.py [hunk 3]
```diff
@@ -0,0 +1,692 @@
...
+    Configuration model for Spanner.
+
+    Attributes:
+        kind (Literal["spanner"]): Type of datastore.
+        project (str): Google Cloud project ID.
+        instance (str): ID of the Spanner instance.
+        database (str): ID of the Spanner database.
+        service_account_key_file (str): Service Account Key File.        
+    """
+    kind: Literal["spanner"]
```
Comment (Yuan325) [inline retrieval_service/datastore/providers/spanner.py:44]: "spanner-gsql" instead? this file can be renamed as `spanner_gsql.py` too.

File: retrieval_service/datastore/providers/spanner.py [hunk 4]
```diff
@@ -0,0 +1,692 @@
...
+from google.cloud.spanner_v1.database import Database
+from google.oauth2 import service_account
+from pydantic import BaseModel
+from google.cloud.spanner_v1 import JsonObject, param_types
+import models
+
+from .. import datastore
+
+# Identifier for Spanner
+SPANNER_IDENTIFIER = "spanner"
```
Comment (Yuan325) [inline retrieval_service/datastore/providers/spanner.py:30]: spanner-gsql instead? since this is specific towards gsql.

**Review by Yuan325 — COMMENTED** [review]
I also ran the tests (there's a lint failure as well, you'll have to run `black .` and `isort .` on "retrieval_service".
Thank you!

File: retrieval_service/models/example-config-spanner.yml [hunk 5]
```diff
@@ -0,0 +1,8 @@
+host: 0.0.0.0
+datastore:
+  # Example for Spanner
+  kind: "spanner"
+  project: "my-project"
+  instance: "my-instance"
+  database: "my_database"
+  service_account_key_file: "path/service/account/credentials.json"
```
Comment (kurtisvg) [inline retrieval_service/models/example-config-spanner.yml:8]: Can we move this into the existing config example? We can include examples of multiple databases and just comment them out until used

**Review by rahul2393 — APPROVED** [review]

File: retrieval_service/models/example-config-spanner.yml (hunk 5 above)
Comment (gauravpurohit06) [reply retrieval_service/models/example-config-spanner.yml:8]: Thanks for the review, currently two example config are present, do we want to merge them into one? 
- retrieval_service/example-config-cloudsql.yml
- retrieval_service/example-config.yml

Comment (kurtisvg) [reply retrieval_service/models/example-config-spanner.yml]: Probably -- I think including the same content just commented out makes the most sense imo

File: retrieval_service/datastore/providers/spanner_gsql.py [hunk 6]
```diff
@@ -0,0 +1,1024 @@
...
+    Configuration model for Spanner.
+
+    Attributes:
+        kind (Literal["spanner"]): Type of datastore.
+        project (str): Google Cloud project ID.
+        instance (str): ID of the Spanner instance.
+        database (str): ID of the Spanner database.
+        service_account_key_file (str): Service Account Key File.
+    """
+
```
Comment (Yuan325) [inline retrieval_service/datastore/providers/spanner_gsql.py:46]:
```suggestion
```

File: retrieval_service/example-config-spanner.yml [hunk 7]
```diff
@@ -0,0 +1,8 @@
+host: 0.0.0.0
+datastore:
+  # Example for Spanner
+  kind: "spanner-gsql"
+  project: "span-cloud-testing"
```
Comment (Yuan325) [inline retrieval_service/example-config-spanner.yml:5]:
```suggestion
  project: "my-project"
  instance: "my-instance"
  database: "my_database"
  service_account_key_file: "my-service-account-key/application_default_credentials.json"
```

**Comment by Yuan325** [comment]
Can we also add a spanner_gsql_test.py for integration test?

**Review by Yuan325 — APPROVED** [review]

**Pull request merged**